### PR TITLE
fix: replace FLOOR() with CAST(AS INTEGER) for SQLite compatibility

### LINKED ISSUE #3102: bug: Search returns server error when using "time of day" filter
### BirdNET-Go version / build date

Current Version: nightly-20260515  Build Date: 2026-05-15T17:02:30Z

### Install method

Docker via install.sh (recommended installer script)

### Device type

Raspberry Pi 5

### Operating system and version

Raspberry Pi OS Trixie Debian 13.4

### Hardware details

4 GB RAM
128 GB microSD card

### Description

On the Search tab, setting the "Time of Day" filter to anything other than "Any Time" returns the error "Search failed: A server error occurred. Please try again later."

### Steps to reproduce

1. Navigate to "Search" tab.
2. Click "Show Advanced Filters" to reveal the "Time of Day" filter.
3. Select anything other than "Any Time" from the "Time of Day" drop-down list.
4. Click "Search"

### Expected behaviour

I would expect the search to return results, filtered by the time of day selected.

### Actual behaviour

Clicking search returns the error "Search failed: A server error occurred. Please try again later."

### Relevant log output

```shell

```

### RTSP source (only if the bug is RTSP related)

_No response_

### Pre-submission checks

- [x] I am running the latest release or have tested with the `:dev` container image.
- [x] I have searched existing open and closed issues and this does not appear to be a duplicate.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

Fixes #3102.

The time-of-day search filter used SQLite's `FLOOR()` math function, which is not available in standard SQLite builds. This caused "no such function: FLOOR" errors when users selected any time-of-day filter other than "Any Time" on the Search tab.

## Changes

- Replace `FLOOR((detected_at + offset) / 3600)` with `CAST((detected_at + offset) / 3600 AS INTEGER)` in the search query builder. `CAST(... AS INTEGER)` is universally supported across SQLite and PostgreSQL.
- Add test `TestSearch_IncludedHoursOnSQLite` covering single-hour, multi-hour, and negative timezone offset filtering on SQLite.

## Testing

- New test exercises `Search()` with `IncludedHours` on an in-memory SQLite database, verifying the query no longer errors.
- All 739 datastore tests pass. Full suite (6494 tests) passes.

_This PR was generated by an automated fix agent based on the technical analysis in #3102. Please review carefully._

## Summary by CodeRabbit

* **Bug Fixes**
  * Hour-based detection filtering is now dialect-aware and correctly computes local "included hours", ensuring consistent results across SQL engines and when applying timezone offsets.

* **Tests**
  * Added test coverage validating hour-based filtering on SQLite, including single- and multi-hour selections and behavior with negative timezone offsets to confirm correct local-hour interpretation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/tphakala/birdnet-go/pull/3107?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: internal/datastore/v2/repository/detection_impl.go [hunk 1]
```diff
@@ -603,9 +603,8 @@ func (r *detectionRepository) buildSearchFilters(query *gorm.DB, filters *Search
 	}
 	if len(filters.IncludedHours) > 0 {
 		// Extract local hour from Unix timestamp using timezone offset.
-		// Formula: FLOOR((detected_at + offset) / 3600) % 24
-		// FLOOR() ensures integer results on both SQLite and MySQL.
-		hourExpr := fmt.Sprintf("FLOOR((detected_at + %d) / 3600) %% 24", filters.TimezoneOffset)
+		// CAST to INTEGER works on both SQLite and PostgreSQL; FLOOR() is unavailable in SQLite.
+		hourExpr := fmt.Sprintf("CAST((detected_at + %d) / 3600 AS INTEGER) %% 24", filters.TimezoneOffset)
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/datastore/v2/repository/detection_impl.go]: Fixed in 6b66ecce - added dialect branch using r.isMySQL: FLOOR() for MySQL, CAST AS INTEGER for SQLite.

File: internal/datastore/v2/repository/detection_impl.go [hunk 2]
```diff
@@ -603,9 +603,12 @@ func (r *detectionRepository) buildSearchFilters(query *gorm.DB, filters *Search
...
 		// Extract local hour from Unix timestamp using timezone offset.
-		// Formula: FLOOR((detected_at + offset) / 3600) % 24
-		// FLOOR() ensures integer results on both SQLite and MySQL.
-		hourExpr := fmt.Sprintf("FLOOR((detected_at + %d) / 3600) %% 24", filters.TimezoneOffset)
+		var hourExpr string
+		if r.isMySQL {
+			hourExpr = fmt.Sprintf("FLOOR((detected_at + %d) / 3600) %% 24", filters.TimezoneOffset)
+		} else {
+			hourExpr = fmt.Sprintf("CAST((detected_at + %d) / 3600 AS INTEGER) %% 24", filters.TimezoneOffset)
+		}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/datastore/v2/repository/detection_impl.go:612]: Table qualification added in dbb56177. Skipping the helper method extraction since the arithmetic-offset approach is fundamentally different from hourFromUnixExpr (which uses SQL date functions for local time).
